### PR TITLE
[SPARK-30115][SQL] Improve limit only query on datasource table

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?

We use Spark as ad-hoc query engine. Most of users' SELECT queries with LIMIT operation like
1) SELECT * FROM TABLE_A LIMIT N
2) SELECT colA FROM TABLE_A LIMIT N
3) CREATE TAB_B as SELECT * FROM TABLE_A LIMIT N
If the TABLE_A is a large table (a RDD with thousands and thousands of partitions), the execution time would be very big since it has to list all files to build a RDD before execution. But almost time, the N is just like 10, 100, 1000, not very big. We don't need to scan all files. This optimization will create a **SinglePartitionReadRDD** to address it.

In our production result, this optimization benefits a lot. The duration time of simple query with LIMIT could reduce 5~10 times. For example, before this optimization, a query on a table which has about one hundred thousands files would run over 30 seconds, after applying this optimization, the time decreased to 5 seconds.

Should support both Spark DataSource Table and Hive Table which can be converted to DataSource table.
Should support bucket table, partition table, normal table.
Should support different file formats like parquet, orc.
This PR only addresses Spark datasource table.
Hive table and view will be filed after this merged.

### How to implement?
1. Add two configurations, `PARTIAL_LISTING_ENABLED` and `PARTIAL_LISTING_MAX_FILES`
2. In `FindDataSourceTable.apply()`, we resolve `GlobalLimit` to add a flag `partialListing` to `FileIndex`
3. In `DataSourceScanExec.inputRDD`, by checking the flag `partialListing` in `relation.location`, we create a `SinglePartitionReadRDD`. This RDD will assign less than `PARTIAL_LISTING_MAX_FILES` files to a single partition.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Add LimitOnlyQuerySuite

### CODE REVIEW

**Comment by LantaoJin** [comment]
cc @mingmwang

**Comment by LantaoJin** [comment]
cc @cloud-fan @dongjoon-hyun @HyukjinKwon

**Comment by cloud-fan** [comment]
The idea looks good, but the implementation looks too complicated. Can we implement it with a single optimizer rule?

**Comment by LantaoJin** [comment]
Emmm, I can try it in another PR. I am afraid it still need many code changes.

**Comment by cloud-fan** [comment]
A hint: AFAIK Hive metastore allows you to fetch up to N partitions, so that we can save time on partitions fetching if there is a limit.

I'm not worried too much about file scan. Spark optimizes LIMIT by `SparkPlan.executeTake`, so we won't scan all the files.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -220,27 +220,47 @@ case class DataSourceAnalysis(conf: SQLConf) extends Rule[LogicalPlan] with Cast
...
+    } else {
+      catalog.getCachedPlan(qualifiedTableName, () => {
+        LogicalRelation(dataSource.resolveRelation(checkFilesExist = false, partialListing), table)
+      })
+    }
   }
 
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case l @ GlobalLimit(_, LocalLimit(limitExpr,
+      Project(projectList, UnresolvedCatalogRelation(tableMeta))))
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:250]: `projectList` should not contain aggregate functions. For example:
```sql
select min(id) from tbl limit 10;
```

**Pull request closed without merging**